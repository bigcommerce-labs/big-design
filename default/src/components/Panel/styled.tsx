import styled from 'styled-components';

import { defaultTheme } from '../../theme';

export const StyledPanel = styled.section`
  ${({ theme }) => theme.elevation.floating}

  background-color: ${({ theme }) => theme.colors.white};
  margin: 0 0 ${({ theme }) => theme.spacing.xxLarge};
  padding: ${({ theme }) => theme.spacing.xxLarge};
  
  &:last-child {
    margin: 0;
  }
`;

StyledPanel.defaultProps = { theme: defaultTheme };
