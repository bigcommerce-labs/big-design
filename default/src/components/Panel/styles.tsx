import { css } from 'styled-components';

export const PanelStyles = css`
  ${({ theme }) => theme.elevation.floating}

  background-color: ${({ theme }) => theme.colors.white};
  margin: 0 0 ${({ theme }) => theme.spacing.xxLarge};
  padding: ${({ theme }) => theme.spacing.xxLarge};
  
  &:last-child {
    margin: 0;
  }
`;
