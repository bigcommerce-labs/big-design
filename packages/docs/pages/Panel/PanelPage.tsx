import { Box, Button, Flex, H2, Panel, Text } from '@bigcommerce/big-design';
import React from 'react';

export default () => (
  <Box marginVertical="xxxLarge" marginHorizontal={{ tablet: 'xxxLarge' }}>
    <Panel>
      <Flex justifyContent="space-between">
        <H2>Panel header - h2</H2>
        <Button variant="secondary">Button</Button>
      </Flex>
      <Text>Text -p</Text>
    </Panel>

    <Panel>
      <H2>Panel header - h2</H2>
      <Text>Text -p</Text>
    </Panel>

    <Panel>
      <Flex justifyContent="space-between">
        <H2>Panel header - h2</H2>
        <Button variant="secondary">Button</Button>
      </Flex>
      <Text>
        Occaecat eiusmod elit laboris pariatur aute esse dolore nostrud ipsum laboris qui. Reprehenderit exercitation
        nisi consectetur deserunt tempor fugiat sunt eiusmod consectetur nisi magna. Consequat aute magna aliquip
        deserunt in magna labore cillum nisi aliquip sit irure anim exercitation.
      </Text>
      <Text>
        Eu occaecat mollit laborum pariatur pariatur excepteur. Voluptate eiusmod ea sit ad deserunt dolore consectetur
        nulla veniam in est aliquip anim adipisicing. Aute sit est mollit commodo qui et anim eu officia proident et.
        Deserunt ipsum et cupidatat magna consectetur aute ad fugiat enim fugiat. Nisi nisi labore anim occaecat non
        aliqua dolor sit nostrud. Do nulla enim proident proident. Fugiat consectetur culpa exercitation veniam culpa
        Lorem irure.
      </Text>
    </Panel>

    <Panel marginBottom="none">
      <H2>Panel header - h2</H2>
      <Text>
        Occaecat eiusmod elit laboris pariatur aute esse dolore nostrud ipsum laboris qui. Reprehenderit exercitation
        nisi consectetur deserunt tempor fugiat sunt eiusmod consectetur nisi magna. Consequat aute magna aliquip
        deserunt in magna labore cillum nisi aliquip sit irure anim exercitation.
      </Text>
      <Text>
        Eu occaecat mollit laborum pariatur pariatur excepteur. Voluptate eiusmod ea sit ad deserunt dolore consectetur
        nulla veniam in est aliquip anim adipisicing. Aute sit est mollit commodo qui et anim eu officia proident et.
        Deserunt ipsum et cupidatat magna consectetur aute ad fugiat enim fugiat. Nisi nisi labore anim occaecat non
        aliqua dolor sit nostrud. Do nulla enim proident proident. Fugiat consectetur culpa exercitation veniam culpa
        Lorem irure.
      </Text>
    </Panel>
  </Box>
);
