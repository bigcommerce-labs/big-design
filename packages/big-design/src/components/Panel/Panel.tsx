import React, { ComponentPropsWithoutRef, forwardRef, isValidElement, memo, Ref } from 'react';

import { MarginProps } from '../../mixins';
import { excludePaddingProps } from '../../mixins/paddings/paddings';
import { warning } from '../../utils';
import { Badge, BadgeProps } from '../Badge/Badge';
import { Button, ButtonProps } from '../Button';
import { Flex } from '../Flex';
import { Text } from '../Typography';

import { StyledH2, StyledPanel } from './styled';

interface PrivateProps {
  forwardedRef: Ref<HTMLDivElement>;
}

export interface PanelAction extends Omit<ButtonProps, 'children'> {
  text?: string;
}

export interface PanelProps extends ComponentPropsWithoutRef<'div'>, MarginProps {
  children?: React.ReactNode;
  description?: React.ReactNode;
  header?: string;
  headerId?: string;
  action?: PanelAction;
  badge?: BadgeProps;
}

export const RawPanel: React.FC<PanelProps & PrivateProps> = memo(({ forwardedRef, ...props }) => {
  const filteredProps = excludePaddingProps(props);
  const { action, children, description, header, headerId, badge, ...rest } = filteredProps;

  const renderHeader = () => {
    if (!header && !action) {
      return null;
    }

    if (typeof header !== 'string') {
      return null;
    }

    return (
      <Flex flexDirection="row">
        {Boolean(header) && (
          <StyledH2 id={headerId} marginBottom={description ? 'xxSmall' : 'medium'}>
            {header}
            {badge && <Badge marginLeft="xSmall" {...badge} />}
          </StyledH2>
        )}
        {action && <Button {...action}>{action.text}</Button>}
      </Flex>
    );
  };

  const renderDescription = () => {
    if (!description) {
      return null;
    }

    if (typeof description === 'string') {
      return <Text color="secondary60">{description}</Text>;
    }

    if (isValidElement(description)) {
      return description;
    }

    warning('description must be either a string or a ReactNode.');
  };

  return (
    <StyledPanel
      marginBottom="medium"
      {...rest}
      backgroundColor="white"
      padding={{ mobile: 'medium', tablet: 'xLarge' }}
      ref={forwardedRef}
      shadow="raised"
    >
      {renderHeader()}
      {renderDescription()}
      {children}
    </StyledPanel>
  );
});

export const Panel = forwardRef<HTMLDivElement, PanelProps>(
  ({ className, style, ...props }, ref) => <RawPanel {...props} forwardedRef={ref} />,
);

Panel.displayName = 'Panel';
