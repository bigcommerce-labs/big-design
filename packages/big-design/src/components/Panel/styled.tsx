import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { StyleableH2 } from '../Typography/private';

export const StyledH2 = styled(StyleableH2)`
  flex-grow: 1;

  & ~ .bd-button {
    width: auto;
    margin-top: 0;
  }
`;

StyledH2.defaultProps = { theme: defaultTheme };
