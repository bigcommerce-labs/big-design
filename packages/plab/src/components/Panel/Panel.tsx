import React from 'react';

import { StyledPanel } from './styled';

export const Panel: React.SFC<{}> = props => <StyledPanel {...props} />;
