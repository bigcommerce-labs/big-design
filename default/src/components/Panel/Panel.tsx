import styled from 'styled-components';

import { defaultTheme } from '../../theme';

import { PanelStyles } from './styles';

export const Panel = styled.section`
  ${PanelStyles};
`;

Panel.defaultProps = { theme: defaultTheme };
