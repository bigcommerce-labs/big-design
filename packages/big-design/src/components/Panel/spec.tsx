import { theme } from '@bigcommerce/big-design-theme';
import React, { createRef } from 'react';
import 'jest-styled-components';

import { fireEvent, render } from '@test/utils';

import { Panel } from './Panel';

test('render panel', () => {
  const { container } = render(
    <Panel>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
      fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit
      labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
    </Panel>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('does not forward styles', () => {
  const { container } = render(
    <Panel className="test" style={{ background: 'red' }}>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
      fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit
      labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
    </Panel>,
  );

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(container.firstChild).not.toHaveStyle('background: red');
});

test('renders a header and action', () => {
  const { getByRole } = render(
    <Panel header="Test Header" action={{ text: 'Test Action' }}>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
      fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit
      labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
    </Panel>,
  );

  const header = getByRole('heading');

  expect(header).toBeInTheDocument();
  expect(header.textContent).toBe('Test Header');

  const actionButton = getByRole('button');

  expect(actionButton).toBeInTheDocument();
  expect(actionButton.textContent).toBe('Test Action');
});

test('action options get forwarded to button', () => {
  const onClick = jest.fn();

  const { getByRole } = render(
    <Panel header="Test Header" action={{ text: 'Test Action', onClick }}>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
    </Panel>,
  );

  fireEvent.click(getByRole('button'));

  expect(onClick).toHaveBeenCalled();
});

test('forwards data attributes', () => {
  const { getByTestId } = render(
    <Panel header="Test Header" data-testid="panel">
      Dolore proident eiusmod sint est enim laboris
    </Panel>,
  );

  const panel = getByTestId('panel');

  expect(panel).toBeInTheDocument();
});

test('forwards ref', () => {
  const ref = createRef<HTMLDivElement>();

  const { container } = render(<Panel ref={ref} />);
  const panel = container.querySelector('div');

  expect(panel).toBe(ref.current);
});

test('ignores padding props', () => {
  const { getByTestId } = render(
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore - ignoring since paddingRight is not a valid prop
    <Panel header="Test Header" data-testid="panel" paddingRight="xxxLarge">
      Dolore proident eiusmod sint est enim laboris
    </Panel>,
  );

  const panel = getByTestId('panel');

  expect(panel).not.toHaveStyle(`padding-right: ${theme.spacing.xxxLarge}`);
});

test("panel action doesn't go to full width", () => {
  const { getByRole } = render(
    <Panel header="Test Header" action={{ text: 'Test Action' }}>
      Test
    </Panel>,
  );

  const button = getByRole('button');

  expect(button).toHaveStyle('width: auto');
});

test('forwards headerId to heading', () => {
  const { getByText } = render(
    <Panel header="Test Header" headerId="test-header">
      Test
    </Panel>,
  );

  expect(getByText('Test Header')).toHaveAttribute('id', 'test-header');
});
