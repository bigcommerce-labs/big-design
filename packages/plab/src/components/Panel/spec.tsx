import 'jest-styled-components';
import React from 'react';
import { render } from 'react-testing-library';

import { Panel } from './Panel';

test('render panel', () => {
  const { container } = render(
    <Panel>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
      fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit
      labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
    </Panel>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('does not forward styles', () => {
  const { container } = render(
    <Panel className="test" style={{ background: 'red' }}>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
      fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit
      labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
    </Panel>,
  );

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(container.firstChild).not.toHaveStyle('background: red');
});
