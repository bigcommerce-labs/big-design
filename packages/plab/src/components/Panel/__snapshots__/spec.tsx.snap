// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render panel 1`] = `
.c0 {
  margin-bottom: 2rem;
  padding: 2rem;
  background-color: #FFFFFF;
  border-radius: 0.25rem;
  box-shadow: 0px 2px 12px rgba(49,52,64,0.2);
}

<div
  class="c0"
  elevation="floating"
>
  Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
</div>
`;
